In [1]:
import pandas as pd
from pathlib import Path
from datetime import datetime

In [2]:
csv_path = Path("./Atlantic_Climate.csv")
df = pd.read_csv(csv_path, low_memory=False)

In [3]:
df.info()
df.head()

<class 'pandas.DataFrame'>
RangeIndex: 964749 entries, 0 to 964748
Data columns (total 17 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Longitude (x)              964749 non-null  float64
 1   Latitude (y)               964749 non-null  float64
 2   Station Name               964749 non-null  str    
 3   Climate ID                 964749 non-null  str    
 4   Date/Time                  964749 non-null  str    
 5   Year                       964749 non-null  int64  
 6   Max Temp (°C)              828481 non-null  float64
 7   Min Temp (°C)              831117 non-null  float64
 8   Mean Temp (°C)             826411 non-null  float64
 9   Heat Deg Days (°C)         826411 non-null  float64
 10  Cool Deg Days (°C)         826411 non-null  float64
 11  Total Rain (mm)            169711 non-null  float64
 12  Total Snow (cm)            165870 non-null  float64
 13  Total Precip (mm)          695482 non-nu

,Longitude (x),Latitude (y),Station Name,Climate ID,Date/Time,Year,Max Temp (°C),Min Temp (°C),Mean Temp (°C),Heat Deg Days (°C),Cool Deg Days (°C),Total Rain (mm),Total Snow (cm),Total Precip (mm),Snow on Grnd (cm),Dir of Max Gust (10s deg),Spd of Max Gust (km/h)
0,-64.51,44.83,AALDERSVILLE,8204020,2021-01-01,2021,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,-64.51,44.83,AALDERSVILLE,8204020,2021-01-02,2021,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,-64.51,44.83,AALDERSVILLE,8204020,2021-01-03,2021,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,-64.51,44.83,AALDERSVILLE,8204020,2021-01-04,2021,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,-64.51,44.83,AALDERSVILLE,8204020,2021-01-05,2021,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [4]:
df.columns

Index(['Longitude (x)', 'Latitude (y)', 'Station Name', 'Climate ID',
       'Date/Time', 'Year', 'Max Temp (°C)', 'Min Temp (°C)', 'Mean Temp (°C)',
       'Heat Deg Days (°C)', 'Cool Deg Days (°C)', 'Total Rain (mm)',
       'Total Snow (cm)', 'Total Precip (mm)', 'Snow on Grnd (cm)',
       'Dir of Max Gust (10s deg)', 'Spd of Max Gust (km/h)'],
      dtype='str')

## Create the three new DataFrames with the specified names

In [5]:
df_wind = df[
    [
        'Climate ID',
        'Date/Time',
        'Dir of Max Gust (10s deg)',
        'Spd of Max Gust (km/h)'
    ]
]

df_station = df[
    [
        'Longitude (x)',
        'Latitude (y)',
        'Station Name',
        'Climate ID'
    ]
]
df_temp = df[
    [
        'Climate ID',
        'Date/Time',
        'Max Temp (°C)',
        'Min Temp (°C)',
        'Mean Temp (°C)',
        'Heat Deg Days (°C)',
        'Cool Deg Days (°C)'
    ]
]

In [6]:
df_temp.info()

<class 'pandas.DataFrame'>
RangeIndex: 964749 entries, 0 to 964748
Data columns (total 7 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Climate ID          964749 non-null  str    
 1   Date/Time           964749 non-null  str    
 2   Max Temp (°C)       828481 non-null  float64
 3   Min Temp (°C)       831117 non-null  float64
 4   Mean Temp (°C)      826411 non-null  float64
 5   Heat Deg Days (°C)  826411 non-null  float64
 6   Cool Deg Days (°C)  826411 non-null  float64
dtypes: float64(5), str(2)
memory usage: 51.5 MB


##  Make df_station contain only unique Climate IDs (EDA for dim_Station)
- To make df_station contain only unique Climate IDs (one row per station), drop duplicate rows based on the Climate ID column.

In [7]:
df_station.info()
df_station = df_station.drop_duplicates(subset='Climate ID').reset_index(drop=True)


<class 'pandas.DataFrame'>
RangeIndex: 964749 entries, 0 to 964748
Data columns (total 4 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Longitude (x)  964749 non-null  float64
 1   Latitude (y)   964749 non-null  float64
 2   Station Name   964749 non-null  str    
 3   Climate ID     964749 non-null  str    
dtypes: float64(2), str(2)
memory usage: 29.4 MB


In [8]:
df_station.head()

,Longitude (x),Latitude (y),Station Name,Climate ID
0,-64.51,44.83,AALDERSVILLE,8204020
1,-54.00,47.31,ARGENTIA (AUT),8400104
2,-67.72,46.71,AROOSTOOK,8100300
3,-65.47,43.45,BACCARO PT,8200255
4,-56.07,48.97,BADGER (AUT),8400301


## EDA for dim_Station 
- compare your station list (df_station) with df_atlantic and then export the cleaned/compared station list to CSV.

In [9]:
df_stationlist = pd.read_csv("./Stationlist_clean.csv")

**Identify Climate IDs NOT matching**
  - Climate IDs in df_station but missing from df_stationlist
  - Climate IDs in df_stationlist but missing from df_station
  - Quick counts (recommended sanity check)

In [10]:
not_in_clean = df_station[
    ~df_station['Climate ID'].isin(df_stationlist['Climate ID'])
]
not_in_station = df_stationlist[
    ~df_stationlist['Climate ID'].isin(df_station['Climate ID'])
]
print("In df_station but not in df_stationlist:",
      not_in_clean['Climate ID'].nunique())

print("In df_stationlist but not in df_station:",
      not_in_station['Climate ID'].nunique())


In df_station but not in df_stationlist: 0
In df_stationlist but not in df_station: 1


**1. Identify the mismatching Climate ID**

In [11]:
missing_ids = df_stationlist.loc[
    ~df_stationlist['Climate ID'].isin(df_station['Climate ID']),
    'Climate ID'
]

missing_ids


69    8502588
Name: Climate ID, dtype: str

**2. Check if this Climate ID exists in temperature data**

In [12]:
missing_climate_id = missing_ids.iloc[0]
df_temp_missing = df_temp[df_temp['Climate ID'] == missing_climate_id]


**3. Inspect the temperature values**

In [13]:
df_temp_missing.info()

<class 'pandas.DataFrame'>
RangeIndex: 0 entries
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Climate ID          0 non-null      str    
 1   Date/Time           0 non-null      str    
 2   Max Temp (°C)       0 non-null      float64
 3   Min Temp (°C)       0 non-null      float64
 4   Mean Temp (°C)      0 non-null      float64
 5   Heat Deg Days (°C)  0 non-null      float64
 6   Cool Deg Days (°C)  0 non-null      float64
dtypes: float64(5), str(2)
memory usage: 132.0 bytes


**Look for it in df_stationlist**

In [14]:
df_stationlist_info = df_stationlist[
    df_stationlist['Climate ID'] == missing_climate_id
]

df_stationlist_info



,Station Name,Province,Latitude,Longitude,Climate ID,First Year,Last Year
69,MAKKOVIK A,NEWFOUNDLAND AND LABRADOR,55.08,-59.19,8502588,2015,2026


# Finalize dim_station_clean

In [15]:
df_station.to_csv("dim_station_clean.csv", index=False)

✅ Final conclusion (important)
This Climate ID:

exists only in Stationlist_clean.csv
has no station metadata in my main dataset
has no temperature records

✅ Therefore, it is invalid for analysis
✅ It should be removed from the station list

## Temperature dim table EDA

In [16]:
df_temp.info()

<class 'pandas.DataFrame'>
RangeIndex: 964749 entries, 0 to 964748
Data columns (total 7 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Climate ID          964749 non-null  str    
 1   Date/Time           964749 non-null  str    
 2   Max Temp (°C)       828481 non-null  float64
 3   Min Temp (°C)       831117 non-null  float64
 4   Mean Temp (°C)      826411 non-null  float64
 5   Heat Deg Days (°C)  826411 non-null  float64
 6   Cool Deg Days (°C)  826411 non-null  float64
dtypes: float64(5), str(2)
memory usage: 51.5 MB


## Count NULL (NaN) values per column

In [17]:
temp_cols = [
    'Max Temp (°C)',
    'Min Temp (°C)',
    'Mean Temp (°C)',
    'Heat Deg Days (°C)',
    'Cool Deg Days (°C)'
]
null_counts = df_temp[temp_cols].isna().sum()
null_counts

Max Temp (°C)         136268
Min Temp (°C)         133632
Mean Temp (°C)        138338
Heat Deg Days (°C)    138338
Cool Deg Days (°C)    138338
dtype: int64

## 1. Find rows where ALL temperature columns are NULL

In [18]:
all_null_rows = df_temp[temp_cols].isna().all(axis=1)
all_null_count = all_null_rows.sum()
all_null_count

np.int64(131560)

## 2. Inspect those rows (important)

In [19]:
df_temp_all_null = df_temp[all_null_rows]
df_temp_all_null.head()
df_temp_all_null.info()

<class 'pandas.DataFrame'>
Index: 131560 entries, 0 to 964577
Data columns (total 7 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Climate ID          131560 non-null  str    
 1   Date/Time           131560 non-null  str    
 2   Max Temp (°C)       0 non-null       float64
 3   Min Temp (°C)       0 non-null       float64
 4   Mean Temp (°C)      0 non-null       float64
 5   Heat Deg Days (°C)  0 non-null       float64
 6   Cool Deg Days (°C)  0 non-null       float64
dtypes: float64(5), str(2)
memory usage: 8.0 MB


## Cleanup "all null" rows

In [20]:
df_temp_clean = df_temp[~all_null_rows].reset_index(drop=True)
df_temp_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 833189 entries, 0 to 833188
Data columns (total 7 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Climate ID          833189 non-null  str    
 1   Date/Time           833189 non-null  str    
 2   Max Temp (°C)       828481 non-null  float64
 3   Min Temp (°C)       831117 non-null  float64
 4   Mean Temp (°C)      826411 non-null  float64
 5   Heat Deg Days (°C)  826411 non-null  float64
 6   Cool Deg Days (°C)  826411 non-null  float64
dtypes: float64(5), str(2)
memory usage: 44.5 MB


**Current situation (based on your output)

df_temp_all_null → ✅ rows where all temp columns were null (already isolated)
df_temp_clean → ✅ all‑null rows removed
Now inside df_temp_clean, there are:

✅ fully complete rows and Partially null rows**

## Find rows with any nulls (after all‑null rows are removed)
 -  Count partially‑null rows

In [21]:
partial_null_rows = df_temp_clean[temp_cols].isna().any(axis=1)
partial_null_count = partial_null_rows.sum()
partial_null_count

np.int64(6782)

In [22]:
df_temp_partial_null = df_temp_clean[partial_null_rows]

df_temp_partial_null.head()

,Climate ID,Date/Time,Max Temp (°C),Min Temp (°C),Mean Temp (°C),Heat Deg Days (°C),Cool Deg Days (°C)
1680,8400104,1995-01-23,NaN,-0.9,NaN,NaN,NaN
1687,8400104,1995-01-30,NaN,-6.2,NaN,NaN,NaN
1689,8400104,1995-02-02,NaN,-5.5,NaN,NaN,NaN
1695,8400104,1995-02-08,NaN,-7.9,NaN,NaN,NaN
1700,8400104,1995-02-13,-4.8,NaN,NaN,NaN,NaN


In [23]:
df_temp_partial_null.info()

<class 'pandas.DataFrame'>
Index: 6782 entries, 1680 to 830316
Data columns (total 7 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Climate ID          6782 non-null   str    
 1   Date/Time           6782 non-null   str    
 2   Max Temp (°C)       2074 non-null   float64
 3   Min Temp (°C)       4710 non-null   float64
 4   Mean Temp (°C)      4 non-null      float64
 5   Heat Deg Days (°C)  4 non-null      float64
 6   Cool Deg Days (°C)  4 non-null      float64
dtypes: float64(5), str(2)
memory usage: 423.9 KB


# NEW RULE DEFINITION FOR ROW VALIDITY 
Remove rows that match this specific invalid pattern:
✅ Heat Deg Days and Cool Deg Days HAVE values
❌ Max Temp, Min Temp, and Mean Temp are ALL null
This is a very reasonable quality rule for climate data.

In [24]:
temp_cols_core = [
    'Max Temp (°C)',
    'Min Temp (°C)',
    'Mean Temp (°C)'
]

degday_cols = [
    'Heat Deg Days (°C)',
    'Cool Deg Days (°C)'
]
invalid_rows = (
    df_temp_clean[temp_cols_core].isna().all(axis=1) &      # all temps missing
    df_temp_clean[degday_cols].notna().any(axis=1)          # but deg days present
)
invalid_rows.sum()

np.int64(0)

# NEW FILLING RULE FOR ROW VALIDITY
✅ Filling rules (restated clearly)
For rows where Mean Temp (°C) is NA:


✅ If BOTH Max and Min are present
→ Mean = (Max + Min) / 2


✅ If ONLY ONE of Max or Min is present
→ Mean = the available value


❌ If BOTH Max and Min are missing
→ Leave Mean as NA (you already removed fully-null rows)

In [25]:
df_t = df_temp_clean.copy()
mask_both = (
    df_t['Mean Temp (°C)'].isna() &
    df_t['Max Temp (°C)'].notna() &
    df_t['Min Temp (°C)'].notna()
)

df_t.loc[mask_both, 'Mean Temp (°C)'] = (
    df_t.loc[mask_both, ['Max Temp (°C)', 'Min Temp (°C)']].mean(axis=1)
)
mask_only_max = (
    df_t['Mean Temp (°C)'].isna() &
    df_t['Max Temp (°C)'].notna() &
    df_t['Min Temp (°C)'].isna()
)

df_t.loc[mask_only_max, 'Mean Temp (°C)'] = df_t.loc[
    mask_only_max, 'Max Temp (°C)'
]
mask_only_min = (
    df_t['Mean Temp (°C)'].isna() &
    df_t['Min Temp (°C)'].notna() &
    df_t['Max Temp (°C)'].isna()
)

df_t.loc[mask_only_min, 'Mean Temp (°C)'] = df_t.loc[
    mask_only_min, 'Min Temp (°C)'
]
df_t['Mean Temp (°C)'].isna().sum()

np.int64(0)

# Spot‑check correctness

In [26]:
df_t.loc[
    mask_both | mask_only_max | mask_only_min,
    ['Max Temp (°C)', 'Min Temp (°C)', 'Mean Temp (°C)']
].head(10)

,Max Temp (°C),Min Temp (°C),Mean Temp (°C)
1680,NaN,-0.9,-0.9
1687,NaN,-6.2,-6.2
1689,NaN,-5.5,-5.5
1695,NaN,-7.9,-7.9
1700,-4.8,NaN,-4.8
1710,NaN,-7.5,-7.5
1712,NaN,-0.2,-0.2
1723,NaN,0.9,0.9
1728,NaN,-6.6,-6.6
1730,NaN,-0.5,-0.5


# Finalize dim_temp data

In [27]:
df_temp_final = df_t
df_temp_final.info()

<class 'pandas.DataFrame'>
RangeIndex: 833189 entries, 0 to 833188
Data columns (total 7 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   Climate ID          833189 non-null  str    
 1   Date/Time           833189 non-null  str    
 2   Max Temp (°C)       828481 non-null  float64
 3   Min Temp (°C)       831117 non-null  float64
 4   Mean Temp (°C)      833189 non-null  float64
 5   Heat Deg Days (°C)  826411 non-null  float64
 6   Cool Deg Days (°C)  826411 non-null  float64
dtypes: float64(5), str(2)
memory usage: 44.5 MB


In [28]:
degday_cols = [
    'Heat Deg Days (°C)',
    'Cool Deg Days (°C)'
]
null_counts = df_temp_final[degday_cols].isna().sum()
null_counts

Heat Deg Days (°C)    6778
Cool Deg Days (°C)    6778
dtype: int64

In [29]:
zero_counts = (df_temp_final[degday_cols] == 0).sum()
zero_counts

Heat Deg Days (°C)     87776
Cool Deg Days (°C)    741738
dtype: int64

In [30]:
(
    (df_temp_final['Heat Deg Days (°C)'] > 0) &
    (df_temp_final['Cool Deg Days (°C)'] > 0)
).sum()


np.int64(0)

In [31]:
df_temp_final.insert(
    0,                      # position (first column)
    'temp_Index',                # column name
    range(1000, 1000 + len(df_temp_final))
)


In [32]:
df_temp_final.info()

<class 'pandas.DataFrame'>
RangeIndex: 833189 entries, 0 to 833188
Data columns (total 8 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   temp_Index          833189 non-null  int64  
 1   Climate ID          833189 non-null  str    
 2   Date/Time           833189 non-null  str    
 3   Max Temp (°C)       828481 non-null  float64
 4   Min Temp (°C)       831117 non-null  float64
 5   Mean Temp (°C)      833189 non-null  float64
 6   Heat Deg Days (°C)  826411 non-null  float64
 7   Cool Deg Days (°C)  826411 non-null  float64
dtypes: float64(5), int64(1), str(2)
memory usage: 50.9 MB


In [33]:
df_temp_final.to_csv("temp_clean.csv", index=False)

In [34]:
df_temp_final.isna().sum()

temp_Index               0
Climate ID               0
Date/Time                0
Max Temp (°C)         4708
Min Temp (°C)         2072
Mean Temp (°C)           0
Heat Deg Days (°C)    6778
Cool Deg Days (°C)    6778
dtype: int64

# EDA for dim_wind

In [35]:
df_wind.info()

<class 'pandas.DataFrame'>
RangeIndex: 964749 entries, 0 to 964748
Data columns (total 4 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Climate ID                 964749 non-null  str    
 1   Date/Time                  964749 non-null  str    
 2   Dir of Max Gust (10s deg)  463946 non-null  float64
 3   Spd of Max Gust (km/h)     469011 non-null  float64
dtypes: float64(2), str(2)
memory usage: 29.4 MB


## Null check

In [36]:

wind_cols = [
    'Dir of Max Gust (10s deg)',
    'Spd of Max Gust (km/h)'
]

df_wind[wind_cols].isna().sum()

Dir of Max Gust (10s deg)    500803
Spd of Max Gust (km/h)       495738
dtype: int64

In [37]:

any_null_rows = df_wind[wind_cols].isna().any(axis=1)
any_null_rows.sum()


np.int64(500803)

In [38]:

both_null_rows = df_wind[wind_cols].isna().all(axis=1)
both_null_rows.sum()


np.int64(495738)

# ✅ How to interpret this (important)


✅ Dir or Speed = NULL usually means:

No gust observed that day
Sensor limitation
Reporting rules (gusts often only logged when above threshold)

✅ Both NULL is very common in climatological wind datasets
→ not an error

⚠️ You should not drop these automatically unless your analysis requires gust data.

## For wind gust data: remove nulls logic

✅ REMOVE rows where both
Dir of Max Gust (10s deg) AND Spd of Max Gust (km/h) are NaN
✅ KEEP rows where at least one of them is present

This matches standard climatological practice.

In [39]:
df_wind_clean = df_wind[~both_null_rows].reset_index(drop=True)
df_wind_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 469011 entries, 0 to 469010
Data columns (total 4 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Climate ID                 469011 non-null  str    
 1   Date/Time                  469011 non-null  str    
 2   Dir of Max Gust (10s deg)  463946 non-null  float64
 3   Spd of Max Gust (km/h)     469011 non-null  float64
dtypes: float64(2), str(2)
memory usage: 14.3 MB


In [40]:
df_wind_clean.head()

,Climate ID,Date/Time,Dir of Max Gust (10s deg),Spd of Max Gust (km/h)
0,8204020,2021-09-11,28.0,43.0
1,8204020,2021-09-12,22.0,49.0
2,8204020,2021-09-13,31.0,34.0
3,8204020,2021-09-15,18.0,43.0
4,8204020,2021-09-16,23.0,33.0


In [41]:

df_wind_clean[
    ['Dir of Max Gust (10s deg)', 'Spd of Max Gust (km/h)']
].isna().all(axis=1).sum()


np.int64(0)

### dim_wind - Columns after cleaning

| Column                         | Non‑null count | Interpretation                                      |
|--------------------------------|----------------|-----------------------------------------------------|
| Climate ID                     | 469,011        | ✅ Intact                                            |
| Date/Time                      | 469,011        | ✅ Intact                                            |
| Dir of Max Gust (10s deg)      | 463,946        | Some gusts lack direction (normal)                  |
| Spd of Max Gust (km/h)         | 469,011        | Every remaining row has gust speed                  |

In [42]:
df_wind_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 469011 entries, 0 to 469010
Data columns (total 4 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Climate ID                 469011 non-null  str    
 1   Date/Time                  469011 non-null  str    
 2   Dir of Max Gust (10s deg)  463946 non-null  float64
 3   Spd of Max Gust (km/h)     469011 non-null  float64
dtypes: float64(2), str(2)
memory usage: 14.3 MB


In [43]:

df_wind_clean.insert(
    0,                      # position (first column)
    'Index',                # column name
    range(100, 100 + len(df_wind_clean))
)


In [44]:
df_wind_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 469011 entries, 0 to 469010
Data columns (total 5 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   Index                      469011 non-null  int64  
 1   Climate ID                 469011 non-null  str    
 2   Date/Time                  469011 non-null  str    
 3   Dir of Max Gust (10s deg)  463946 non-null  float64
 4   Spd of Max Gust (km/h)     469011 non-null  float64
dtypes: float64(2), int64(1), str(2)
memory usage: 17.9 MB


In [45]:
df_wind_clean.to_csv("dim_wind_clean.csv", index=False)